### ISSUE TITLE
Glassfish start-domain fails after waiting for long

### ISSUE THREAD

**Issue by alwin-joseph** [eclipse-ee4j/glassfish#23191]
## Environment Details
* GlassFish Version (and build number): 
https://download.eclipse.org/ee4j/glassfish/glassfish-6.0.0-SNAPSHOT-nightly.zip
https://download.eclipse.org/ee4j/glassfish/glassfish-6.0.0-SNAPSHOT-nightly.info 
e4ef2bf0c5b9bf732c1b1393510f7c11919bd4064f606f864f07193f3cb9af6e  
-rw-r--r-- 1 jenkins 1001880000 114124597 Aug 26 04:59

* JDK version:  8u191
* OS: Linux
* Database: Derby

----------

## Problem Description
asadmin start-domain is waiting for a long time(600 seconds)
"Waiting for domain1 to start"

Issue is intermittent.

I encountered the same in own virtual machine.  
verified jps -l, ps -eaf commands to check any running processes but found nothing.
Also there does not seem to be any memory issues as far as I see.

logs of start-domain with verbose on : 
[verboselogs.txt](https://github.com/eclipse-ee4j/glassfish/files/5130976/verboselogs.txt)

threaddump (ctrl+ \ )
[threaddump.txt](https://github.com/eclipse-ee4j/glassfish/files/5130978/threaddump.txt)

server.log
[server.log](https://github.com/eclipse-ee4j/glassfish/files/5131119/server.log)

It would be great if someone can go through the logs to identify any issue at glassfish.

The verbose logs suggested  below : 

Caused by: java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: Lorg/glassfish/flashlight/MonitoringRuntimeDataRegistry;

Caused by: java.lang.ClassNotFoundException: org.glassfish.flashlight.MonitoringRuntimeDataRegistry not found by org.glassfish.main.admin.monitoring-core [190]

**Comment by scottmarlow** [comment]
I can recreate this on my local machine (running Fedora 32).  There wasn't a resolution when we discussed on https://www.eclipse.org/lists/glassfish-dev/msg00502.html but good to hear that it also recreates for others.

[I have been running GlassFish 6.0 under Podman (```podman run -it centos /bin/bash```) as a workaround](https://github.com/eclipse-ee4j/jakartaee-tck/wiki/Instructions-for-running-the-nightly-engineering-build-of-Platform-TCK-with-Podman-(currently-for-EE-9)).

**Comment by Cousjava** [comment]
This occurs for me with both locally built master and downloaded nightly snapshot.

**Comment by scottmarlow** [comment]
@Cousjava which operating system are you using?  As mentioned above I'm using Fedora 32 locally.

**Comment by Cousjava** [comment]
I'm using Ubuntu MATE 20.04. using Zulu JDK 8 build 1.8.0_262-b18.

**Comment by scottmarlow** [comment]
I set a breakpoint in org.jvnet.hk2.internal.SystemDescriptor.reify at line 684 which does `collector.addThrowable(new IllegalArgumentException("Errors were discovered while reifying " + this));` and step up through the code which returns back to com.sun.enterprise.v3.server.AppServerStartup.run() line 292 which returns without an indication of failure, we return to AppServerStartup.doStart() just after the call to run(), which then creates a new CountDownload latch and starts a non-daemon thread that waits indefinitely for shutdown request.` which is ignoring the failure and after latch is created we see in console `Server shutdown initiated` which is ignored as shutdownSignal is still false, so instead we just keep the server process in wedged state it seems.

At the very least, I think we should have void run() return true or false if server should be shutdown.  The question is, should the server shutdown on the `java.lang.ClassNotFoundException: org.glassfish.flashlight.MonitoringRuntimeDataRegistry` failure or should we instead ignore that and continue starting up the server, it should be one or the other.

In summary, there are two issues here:

1. Server startup failures that are important should cause the server to terminate instead of leaving server in partially started state.
2.  The java.lang.ClassNotFoundException failure should be fixed.

**Comment by scottmarlow** [comment]
My guess is that the [classpath ](https://gist.github.com/scottmarlow/0e0ce78a9da4519b416a63490eecdd06)is missing a jar dependency that is also needed.

**Comment by joeyleeeeeee97** [comment]
Any updates on this?
I am trying with jakartaee-tck Jenkins pipeline and also met this problem. Pretty much the same exception stack and verbose log.
http://ci.dragonwell-jdk.io/job/jakartaee-tck/33/console

**Comment by joeyleeeeeee97** [comment]
Appreciate if there is any workaround based on the current Jenkins job that I could try. :)

**Comment by Cousjava** [comment]
Workaround that seems to do the trick for me:

First, I moved monitoring-core.jar out of the modules directory so it wouldn't be picked up by GlassFish. Then I ran start-domain, which has the server fails to start because other modules depend on monitoring-core. I then moved monitoring-core.jar back into the modules directory and then GlassFish worked fine.

**Comment by joeyleeeeeee97** [comment]
@Cousjava .  It worked, Thanks for your help.

**Comment by scottmarlow** [comment]
The workaround works for me locally also!  

I'm now thinking about how I would do this in various places in the Jakarta EE Platform TCK CI scripting to ensure this doesn't happen.

Should we have a GlassFish 6.0 script for doing this or perhaps include some extra files for osgi-cache that simulate what this does?

CC @smillidge

**Comment by smillidge** [comment]
Problem is I'm not sure what this workaround is actually doing and whether a GlassFish script helps. I can see that there could be a TCK script that does this as it starts and stops GlassFish a number of ties when building the domain.

I'll see if @Cousjava can pin this down on the GlassFish side as he can reproduce.

**Comment by scottmarlow** [comment]
I compared the difference between the contents of the osgi-cache after the first start of GlassFish that is missing the monitoring-core.jar versus the second start of GlassFish which contains the monitoring-core.jar jar.

The only difference that I see in osgi-cache is:
1. glassfish/domains/domain1/osgi-cache/felix/bundle0/bundle.id contains `311` after the first GlassFish run
1. glassfish/domains/domain1/osgi-cache/felix/bundle0/bundle.id contains `313` after the second GlassFish run.

I assume that the bundle.id is simply assigned and not that meaningful, so I agree that this could be worked around by users I guess since we don't know what the cause is exactly.

**Comment by arjantijms** [comment]
> I assume that the bundle.id is simply assigned and not that meaningful

From practical experience and from memory I think these are normally pretty constant if not changes, even though they are dynamically assigned in the order Felix puts modules into service.

**Comment by smillidge** [comment]
Has anybody tried `-Dorg.glassfish.startupThreadPolicy=USE_NO_THREADS`

**Comment by scottmarlow** [comment]
I think that I tried that before, I'll try `-Dorg.glassfish.startupThreadPolicy=USE_NO_THREADS` now in config/domain.xml:
```
 <jvm-options>-Dorg.glassfish.startupThreadPolicy=USE_NO_THREADS</jvm-options>
```
Still got the `java.lang.NoClassDefFoundError: Lorg/glassfish/flashlight/MonitoringRuntimeDataRegistry`

**Comment by smillidge** [comment]
@scottmarlow I've tried to change the order of the boot in the monitoring registry. I can't reproduce the issue but this pushed back in the boot order when the problematic class needs to be resolved.

**Comment by Cousjava** [comment]
No change, I still get the same error.

**Comment by scottmarlow** [comment]
@smillidge I downloaded and tried glassfish.zip from [job/PR-23258](https://ci.eclipse.org/glassfish/view/GlassFish/job/glassfish_build-and-test-using-jenkinsfile/view/change-requests/job/PR-23258) but still got the same failure (see attached server.log that contains `Caused by: java.lang.ClassNotFoundException: org.glassfish.flashlight.MonitoringRuntimeDataRegistry not found by org.glassfish.main.admin.monitoring-core`).

[server.log](https://github.com/eclipse-ee4j/glassfish/files/5438887/server.log)

What is the best way to explore the cause?  Debugging a certain class or perhaps adding print/log output to certain classes?  The failure occurs nearly 100% of the time for me locally.

**Comment by smillidge** [comment]
@scottmarlow attached server log doesn't have any exception in it. Do you have a different one?

**Comment by scottmarlow** [comment]
Sorry about that, attached here is the [console output which does have the exception.](https://github.com/eclipse-ee4j/glassfish/files/5439906/glassfish_console_pr-23258.txt)

**Comment by Cousjava** [comment]
I have created a PR which fixed the issue for locally, I've rebuilt GlassFish a couple of times with this change and tried starting a domain successfully.

**Comment by smillidge** [comment]
@scottmarlow can you try the zip built from @Cousjava's PR https://ci.eclipse.org/glassfish/job/glassfish_build-and-test-using-jenkinsfile/job/PR-23264/1/ and see if that fixes it for you.

**Comment by scottmarlow** [comment]
@smillidge yes, I verified that the zip built from https://ci.eclipse.org/glassfish/job/glassfish_build-and-test-using-jenkinsfile/job/PR-23264/1/ did work for me locally (e.g. just starting GlassFish but not running any TCK tests).  

I then downloaded https://download.eclipse.org/ee4j/glassfish/glassfish-6.0.0-SNAPSHOT-nightly.zip and verified that I can still reproduce the `ClassNotFoundException: org.glassfish.flashlight.MonitoringRuntimeDataRegistry not found by org.glassfish.main.admin.monitoring-core ` locally.  So it looks good.

**Comment by smillidge** [comment]
OK I will merge in @Cousjava's PR as it seems to help

**Comment by smillidge** [comment]
Hopefully fixed by merge of #23264

**Comment by scottmarlow** [comment]
> Hopefully fixed by merge of #23264

Wow, our first perfect TCK run with #23264!!!
https://ci.eclipse.org/jakartaee-tck/job/jakartaee-tck-scottmarlow/job/darranl-unauthorized_status_code/2/testReport

Thanks @Cousjava for addressing this issue, very much appreciated!

**Comment by gerdogdu** [comment]
> @smillidge I downloaded and tried glassfish.zip from [job/PR-23258](https://ci.eclipse.org/glassfish/view/GlassFish/job/glassfish_build-and-test-using-jenkinsfile/view/change-requests/job/PR-23258) but still got the same failure (see attached server.log that contains `Caused by: java.lang.ClassNotFoundException: org.glassfish.flashlight.MonitoringRuntimeDataRegistry not found by org.glassfish.main.admin.monitoring-core`).
> 
> [server.log](https://github.com/eclipse-ee4j/glassfish/files/5438887/server.log)
> 
> What is the best way to explore the cause? Debugging a certain class or perhaps adding print/log output to certain classes? The failure occurs nearly 100% of the time for me locally.

For me, this glassfish.zip started without any problem in my macOS Catalina 10.15.6 with 

openjdk version "1.8.0_252"
OpenJDK Runtime Environment (build 1.8.0_252-b09)
OpenJDK 64-Bit Server VM (build 25.252-b09, mixed mode)

Regards.
Gurkan

**Comment by gerdogdu** [comment]
> > Hopefully fixed by merge of #23264
> 
> Wow, our first perfect TCK run with #23264!!!
> https://ci.eclipse.org/jakartaee-tck/job/jakartaee-tck-scottmarlow/job/darranl-unauthorized_status_code/2/testReport
> 
> Thanks @Cousjava for addressing this issue, very much appreciated!

Great :)
